### PR TITLE
using route scoring on all route matches, returning the highest score

### PR DESCRIPTION
**Pull request by stackoverfloweth**
add concept of route score: more expected query params fulfilled the higher the value. If equal then the route with higher depth will have the higher value

updated route match logic to get score for each potential match, return first (sort desc, so first with highest score)

### CODE REVIEW

File: src/utilities/routeMatchScore.spec.ts [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+    const [aRoute, bRoute] = resolveRoutes(routes)
+
+    const sortByRouteScore = getRouteScoreSortMethod('/?color=red&id=1&extra=ok')
+    const response = [aRoute, bRoute].sort(sortByRouteScore)
+
+    expect(response).toMatchObject([bRoute, aRoute])
+  })
+
+  test('given routes with different query scores and asc sort, returns them sorted by query score ascending', () => {
```
Comment (pleek91) [inline src/utilities/routeMatchScore.spec.ts:91]: Curious your thoughts on the sorts. What do we need asc for?

**Review by pleek91 — COMMENTED** [review]
Couple questions but overall this makes sense to me, nicely done 👍

File: src/utilities/routeMatchScore.spec.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    } as const satisfies Route
+    const [resolved] = resolveRoutes([route])
+    const actualQuery = new URLSearchParams('/?can=have&some=queries')
+
+    const response = countExpectedQueryKeys(resolved, actualQuery)
+
+    expect(response).toBe(0)
+  })
+
+  test('given route with query where url has keys but different values, still counts as present', () => {
```
[the remark this answers is not in the document]
Comment (stackoverfloweth) [reply src/utilities/routeMatchScore.spec.ts:20]: you're correct. And I recognized that such behavior was not obvious why we're only matching keys and note values, so to protect ourselves from future us coming back in here and adding unnecessary complexity I wrote a test.

File: src/utilities/routeMatchScore.spec.ts (hunk 1 above)
Comment (stackoverfloweth) [reply src/utilities/routeMatchScore.spec.ts:91]: over-engineering for no good reason

Comment (pleek91) [reply src/utilities/routeMatchScore.spec.ts:91]: So should we remove the sorts or do you think they'll have value in the future

Comment (stackoverfloweth) [reply src/utilities/routeMatchScore.spec.ts]: removed 👍

**Review by pleek91 — APPROVED** [review]

**Pull request merged**